### PR TITLE
Fix bug in pycbc_inspiral

### PR DESCRIPTION
**Pull request by spxiwh**
We've been tracking down a nasty issue in pycbc_inspiral recently. The symptom we saw was a BNS trigger with SNR ~ 12 and chisq ~ 1 per DOF. Except all the SNR was coming from a really nasty glitch *right* at the start of the waveform, for which we expect chisq to be large. Running this through `pycbc_single_template` found the expected large chisq.

We expected some problem in the chisq code, or in the GPU code, or in the new compressed waveforms, which meant it took us a long time to find the *actual* problem.

There's a bug in the sigmasq calculation for SPATmplt ... Or at least in the caching of it. We are not checking if the *lower* frequency has changed. So if the first template has a f_lower of 25Hz, all subsequent templates not using that f_lower will be affected.

I suspect this has been in place for some time now (since the start of O2?). So this will be a factor in at least GWTC-1 results, and 2-ODC (and 1-ODC?). However, it looks like it's never really made a major difference until this one example, so I'm hoping this is just a weird edge-case that really emphasizes this bug! The SNR variation due to this seems small (< 1%). I'm guessing it's just very strange cases where the chisq makes a difference.

I have tested that this fixes the issue in some small examples. Might be worth also testing it in some larger jobs that Josh has handy.

@ahnitz Does this seem like the best way of resolving this? Are the memory ramifications to doing this?

### CODE REVIEW

**Comment by tdent** [comment]
We're supposing that a sigmasq vector starting from the 'wrong' frequency for a given template has some quite bad effects on the chisq calculation, but we haven't tracked that down explicitly. 
It _might_ be worth considering whether chisq calculation should fail (throw an error) under these circumstances, but we also want to avoid false positives.

**Comment by spxiwh** [comment]
@tdent I've just pushed a second patch to hook up some things in chisq to use the new cached arrays. I'm not sure additional checks are required on top of this (these are optimized paths, so we don't want to slow them down). A "standard" user trying chi-squared would not use such paths unless they do out of their way to do so.

**Comment by josh-willis** [comment]
I'll go ahead and add my approval, because I think this correctly fixes the bug we are seeing, but I want to highlight another issue in this same part of the code, and ask @ahnitz a question. They are not the problem that is causing our immediate bug, but there is code here whose benefit is at least unclear to me, and also possibly risky.

First, a typo, that has the effect of disabling  a feature, but I don't know how important that feature is to us. In [this line](https://github.com/gwastro/pycbc/blob/master/pycbc/vetoes/chisq.py#L332), we create the per-template cache of bins as an instance of our own class, `LimitedSizeDict`. However instead of supplying the keyword argument `size_limit`, that line says `size_limite`, which doesn't exist, meaning the class will be created with no size limit. Why are we imposing a size limit, and why is it four? What happens given that that limit in fact isn't being enforced?

More importantly, I'd like to understand, without tracking through too much code, exactly what is being cached here and why.  We are creating a cache for each psd, keyed by the id of the template params (which is unsafe, as `id(object)` is not guaranteed to be unique over the life of an executable). The value for that key in that dict is then set to be `True` on line 335. What is that recording; namely, why is this just a boolean? Is there a case where the key is in the dict, but the value is `False`?

Secondly, the code block on [lines 335--345](https://github.com/gwastro/pycbc/blob/master/pycbc/vetoes/chisq.py#L335-L345) is controlled by this `if` statement: `key not in template._bin_cache or id(template.params) not in psd._chisq_cached_key`. I really don't understand the second clause (after `or`) given what we see, but this goes back to my earlier question about what the per-psd binary-valued cache is recording.  I *believe* I am more clear about the per-template cache of the bins themselves, which is created if we haven't encountered that template before with that PSD.  Here it clearly makes a performance difference to avoid recalculating the `sigmasq` vector when we can (but this is why Ian's fix is needed), but why are we caching the bins themselves, since I think they will only be reused if that exact template is reused with that exact PSD. Am I correct in that analysis, because if so, that doesn't seem like a very common occurrence?

More generally, whatever I am misunderstanding, we are using `id` as a key which isn't safe (and has bitten us in the past) and we seem to be doing some sort of double-caching.  If there is indeed a use case for that, we need to add comments to this code to explain what is going on.

**Review by josh-willis — APPROVED** [review]

**Comment by ahnitz** [comment]
> I'll go ahead and add my approval, because I think this correctly fixes the bug we are seeing, but I want to highlight another issue in this same part of the code, and ask @ahnitz a question. They are not the problem that is causing our immediate bug, but there is code here whose benefit is at least unclear to me, and also possibly risky.
> 
> First, a typo, that has the effect of disabling a feature, but I don't know how important that feature is to us. In [this line](https://github.com/gwastro/pycbc/blob/master/pycbc/vetoes/chisq.py#L332), we create the per-template cache of bins as an instance of our own class, `LimitedSizeDict`. However instead of supplying the keyword argument `size_limit`, that line says `size_limite`, which doesn't exist, meaning the class will be created with no size limit. Why are we imposing a size limit, and why is it four? What happens given that that limit in fact isn't being enforced?
> 
It will use some extra memory. 

> More importantly, I'd like to understand, without tracking through too much code, exactly what is being cached here and why. We are creating a cache for each psd, keyed by the id of the template params (which is unsafe, as `id(object)` is not guaranteed to be unique over the life of an executable). The value for that key in that dict is then set to be `True` on line 335. What is that recording; namely, why is this just a boolean? Is there a case where the key is in the dict, but the value is `False`?

No, it was just a dummy value so I could test for membership in a dict. It is never False.

> Secondly, the code block on [lines 335--345](https://github.com/gwastro/pycbc/blob/master/pycbc/vetoes/chisq.py#L335-L345) is controlled by this `if` statement: `key not in template._bin_cache or id(template.params) not in psd._chisq_cached_key`. I really don't understand the second clause (after `or`) given what we see, but this goes back to my earlier question about what the per-psd binary-valued cache is recording. I _believe_ I am more clear about the per-template cache of the bins themselves, which is created if we haven't encountered that template before with that PSD. Here it clearly makes a performance difference to avoid recalculating the `sigmasq` vector when we can (but this is why Ian's fix is needed), but why are we caching the bins themselves, since I think they will only be reused if that exact template is reused with that exact PSD. Am I correct in that analysis, because if so, that doesn't seem like a very common occurrence?

The thing to realize is that if the id of PSD has changed, it means it has either been regenerated, *or* is a different object that came back with the same id value. In both cases it will no longer have the cache attached to the PSD. So the cache-in-psd + id of psd guarantees uniqueness. That is why it checks both the template cache and that *this* PSD really was the one the cache was generated for (even if the id was reused). 

TLDR, the 'or' is what exactly protects against the facts that ids can be reused. 


> 
> More generally, whatever I am misunderstanding, we are using `id` as a key which isn't safe (and has bitten us in the past) and we seem to be doing some sort of double-caching. If there is indeed a use case for that, we need to add comments to this code to explain what is going on.

File: pycbc/vetoes/chisq.py [hunk 1]
```diff
@@ -335,11 +335,16 @@ def cached_chisq_bins(self, template, psd):
             psd._chisq_cached_key[id(template.params)] = True
             num_bins = int(self.parse_option(template, self.num_bins))
```
Comment (ahnitz) [inline pycbc/vetoes/chisq.py:336]: I won't hold this up if you want to merge it urgently, but I don't believe this is the best solution. 

The issue would seem to be thta the sigmas_vec cache itself was not generated with a low enough flow to cover the range of flows that might be encountered. The solution would seem to be to generate the cache from the lowest flow in the bank (it can know this). 

The solution here adds some extra complexity but also causes the cache to be regenerated for every template + PSD combo (unless some templates have flow the same). The only other saving would be if some data segments use the same PSD, but that also is not used to a high degree.

Comment (ahnitz) [reply pycbc/vetoes/chisq.py:336]: I think the better fix would be to use the bank flow here (self.min_f_lower)
https://github.com/gwastro/pycbc/blob/master/pycbc/waveform/bank.py#L63

and then take a difference of kmin / kmax here instead of just kmax
https://github.com/gwastro/pycbc/blob/master/pycbc/waveform/bank.py#L74

Comment (spxiwh) [reply pycbc/vetoes/chisq.py:336]: Agreed. I've changed the patch to do this now.

(As the templates have lower frequencies rounded to the nearest 0.1Hz, there's not a large number (O(10)) of different low-frequency cutoffs used for the SPATmplt templates, but I agree it's better not to compute the same thing multiple times.)

**Review by ahnitz — APPROVED** [review]
This looks correct to me. I'm assuming you've verified in hand testing though.

**Comment by spxiwh** [comment]
I've tested that I'm getting the same output before/after applying the additional change in this patch (and that in both cases we get the same output as when only filtering a single template).
